### PR TITLE
Implement a node counter which sets a `data-node-id` to every component node

### PR DESCRIPTION
**Pull request by janmarkuslanger**

### CODE REVIEW

File: src/core/createElement/createElement.js [hunk 1]
```diff
@@ -69,10 +72,16 @@ const appendElementVnode = (element, vnode) => {
...
  */
-export const createElement = (vnode = {}) => {
+export const createElement = (vnode = {}, Component) => {
   // create the element
   const element = document.createElement(vnode.tagName);
 
+  // if a component gets passed asign and increment node id
+  if (Component) {
+    Component.nodeCounter += 1;
+    vnode.props[DATA_NODE_ATTRIBUTE] = Component.nodeCounter;
```
Comment (dysfunc) [inline src/core/createElement/createElement.js]: this is missing 

element.dataset[DATA_NODE_ATTRIBUTE] = Component.nodeCounter;

File: src/core/constants.js [hunk 2]
```diff
@@ -11,3 +11,10 @@ export const DATA_COMPONENT_ATTRIBUTE = 'data-power-component';
  * @type {Boolean}
  */
 export const IS_COMPONENT = 'IS_COMPONENT';
+
+/**
+ * data node id
+ * @private
+ * @type {Number}
+ */
+export const DATA_NODE_ATTRIBUTE = 'data-node-id';
```
[the remark this answers is not in the document]
Comment (janmarkuslanger) [reply src/core/constants.js]: Then i will change `data-power-component` to `power-component`too.

File: src/core/component/component.js [hunk 3]
```diff
@@ -69,7 +72,7 @@ export class Component {
     this.componentVdom = this.render();
 
     // get the template by call the render
-    this.template = createElement(this.componentVdom);
+    this.template = createElement(this.componentVdom, this);
```
[the remark this answers is not in the document]
Comment (janmarkuslanger) [reply src/core/component/component.js]: How do we get the component to its children? 
```
const appendElementVnode = (element, vnode, Component) => {
  element.appendChild(createElement(vnode, Component));
};
```

[the remark this answers is not in the document]
Comment (dysfunc) [reply src/core/component/component.js]: Like this:

```js
export const createElement = (Component) => {
  const vnode = Component.componentVdom || {};
  const { tagName = 'div', props = {} } = vnode;
  
  // create the element
  const element = document.createElement(vnode.tagName);

  // if a component gets passed asign and increment node id
  if (Component) {
    Component.nodeCounter += 1;
    vnode.props[DATA_NODE_ATTRIBUTE] = Component.nodeCounter;
  }

  if (isObject(vnode.props)) {
    appendElementObject(element, vnode.props);
  }

  vnode.children.forEach((child) => {
    if (isString(child)) {
      appendElementText(element, child);
    } else if (isVnode(child)) {
      if (Component) {
        appendElementVnode(element, child, Component);
      } else {
        appendElementVnode(element, child);
      }
    }
  });

  return element;
};
```

File: src/core/constants.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (dysfunc) [reply src/core/constants.js]: +1

**Pull request merged**